# The apps most people want

The project consist in analyze all game apps that exists in Plays Store and iOS store so we can see which type of app games are the one the people most like.

Our goal is to have enough information of the interest of the people in app games so our developers can create games as their interest.

In [1]:
import csv 
op=open('AppleStore.csv')
op2=open('googleplaystore.csv')
read1=csv.reader(op)
read2=csv.reader(op2)
lista1=list(read1)
lista2=list(read2)#Google Play
nuevo1=lista1[1:]
nuevo2=lista2[1:]

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [26]:
explore_data(nuevo1,0,1,True)
explore_data(nuevo2,0,1,True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


Number of rows: 7197
Number of columns: 16
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


Number of rows: 10840
Number of columns: 13


## Number of rows and columns in Apple Store data Base

In [4]:
number_row=len(nuevo1)
number_column=len(nuevo1[0])
print(number_row, number_column)

7197 16


## Number of rows and columns in Google Play Store

In [5]:
number_row2=len(nuevo2)
number_column2=len(nuevo2[0])
print(number_row2, number_column2)

10841 13


## Column Names in each data set

In [6]:
HeaderAp=lista1[0]
HeaderGp=lista2[0]
print(HeaderAp)#Column names of Apple Store
print('\n')
print(HeaderGp)#Column names of Google Play

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


The columns that could help us in our goal are:

For Apple Store:
'prime_genre', 
'user_rating', 
'price'

For Google Play:
'Category', 
'Rating', 
'Installs', 
'Type'

If you get problems to figure it out what each column is about, here your have the documentation:

Apple Store Data Set:  [Link](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)

Google Play Data Set: [Link](https://www.kaggle.com/lava18/google-play-store-apps)


In [7]:
print(nuevo2[10472])
print(len(nuevo2[10472]))
print(len(HeaderGp))

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
12
13


In [8]:
longi=len(HeaderGp)
for each in nuevo2:
    lon=len(each)
    if lon!=longi:
        print(each)
        index=nuevo2.index(each)
        del nuevo2[index]

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [9]:
print(len(nuevo2))

10840


## Removing Duplicates

Deleting apps with missing values is one step of cleaning data. Another thing we need to consider is if it exists duplicate values. For example, if the name of an app is repeated more than once. 

In order to have a good analysis, we need to work only with one app specification.

As an example, in the next code we will print all the rows that has 'Instagram' as a name app.

In [10]:
for row in nuevo2:
    name=row[0]
    if name=='Instagram':
        print(row)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


Here we can see that 'Instagram' has 4 entries. We only need to work with one of these, so we can take a value as the cretreria to decide which are the ones to be deleted.

If we want to check how many apps are duplicated we run the next code:

In [11]:
unique_apps=[]
duplicated_apps=[]
for each in nuevo2:
    name=each[0]
    if name in unique_apps:
        duplicated_apps.append(name)
    else:
        unique_apps.append(name)
print('Unique apps:', len(unique_apps))
print('/n')
print('Duplicated apps:', len(duplicated_apps))


Unique apps: 9659
/n
Duplicated apps: 1181


It was mentioned before that we would use a value as criteria in order to choose which row will remain.

We can delete ramdomly the rows, but doing that we can loose valuable information due an update of the information, and update of the app. So what we do is to check the information af all the rows of a repetitive name. Let's take Instagram as example:

If we analize in detail we can see that all values are the same, like rating or number od downloads, type or even the last Update. The only difference here is the number of reviews. We can take this as the criteria choosing the one with higher value.

Ok, so we want to separate the duplicated values from the rest:

First we will creat an empty dictionarie called `reviews_max`
  -  Then we iterate every `row` in `nuevo2` the database:
     -  For every row we asign its element in index 0, which              has the app name to a variable called `name`.
     -  For every row we asign its element in index 3, which              has the reviews to a variable called `n-reviews`
     
Then with those tho values we creat a conditional, where we 
check if the name in the current row is inside the               dictionarie `reviews_max`.
  
     
  -  If it is then we pass into another conditional:
        -  where              checks if the actual value of the rating, which is                stored in `n_reviews` is bigger than the value of the            key stored inside the dictionarie. If it is, then we              change the value.*
  
If the name is not inside the dictionarie, then we add it 

And with that we have in our dictionarie as a key the name of the app and as the value the bigger rating that has asigned to tah name. So in other words, this dictionarie has the information of the rows in the data base that we need.
     

In [12]:
'''Aqui encontramos las apps con la info que deben 
tener para ser agregadas'''
reviews_max={}
for row in nuevo2:
    name=row[0]
    n_reviews=float(row[3])
    if name in reviews_max:
        if n_reviews>reviews_max[name]:
            reviews_max[name]=n_reviews
    else:
        reviews_max[name]=n_reviews
print(len(reviews_max))

9659


Now its time to get the rows we need.

- So first we create two empty lists called `android_clean` and 
`already_added`.

- Then we iterate for every row in our database:
    - In the current row at index number 0, which has the app name, we asign it to a variable called `name`.
    -In the current row at index number 3, which has the app reviews, we asign it to a variable called `n_reviews`.
    
Now, it is important to have in mind the dictionarie that we created before called `reviews_max`.

- Wewill make an if statement where it will check the following:
    - The current reviews are the same as the one we have in the ditionarie for that name, that in the dictionarie es the key.
    - The current name is not inside the `already_added` list. This is very important, because we may have, in the duplicates, the same higher review value for more than one entry , so, if that is the case, then we will add duplicates inside our `android_clean` list.
    
    - Finally we add the row inside the `android_clean` list and the name insde the `already_added` list.

In [13]:
android_clean=[]
already_added=[]
for row in nuevo2:
    name=row[0]
    n_reviews=float(row[3])
    if name not in already_added and n_reviews==reviews_max[name]:
        android_clean.append(row)
        already_added.append(name)
print(len(android_clean))
        
        
        
    

9659


In [14]:
def english(string):
    i=0
    for letter in string:
        num=ord(letter)
        if num>127:
            i+=1
    if i>3:
        return False       
    return True

print(english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

False


In [27]:
#Checamos cuantas apps pueden no estar en ingles, si lo estan
#lo agregamos a otra lista Apple
lang_en1=[]
for row in nuevo1:
    name=row[1]
    language=english(name)
    if language==True:
        lang_en1.append(row)
print(len(lang_en1))
        
    

6183


In [16]:
#Checamos cuantas apps pueden no estar en ingles, si lo estan
#lo agregamos a otra lista Play Store
lang_en2=[]
for row in android_clean:
    name=row[0]
    language=english(name)
    if language==True:
        lang_en2.append(row)
print(len(lang_en2))

9614


## Cleaning Non Free Apps

In [28]:
#Google Play Store
final2=[]
for row in lang_en2:
    type=row[7]
    if type=='0':
        final2.append(row)
print(len(final2))


8864


In [29]:
#para iOS
final1=[]
for row in lang_en1:
    type=float(row[4])
    if type==0.0:
        final1.append(row)
print(len(final1))
    

3222


## Starting the analysis

The revenue of the company is influenced by the number of users in our apps due the apps that we create are free to use, so the income comes from publicity. The more the poeple use our apps the more money we get.

Create an app consumes a lot of resources, to minimize risks and losts, the creating of an app follows the next steps:

- Build a minimal Anfroid Version of the app and add it to Google Play Store.
- If the app has a good response, the app is developed further.
- If the app is profitable after six months, we built an iOs versions of the app.

We wan that the app has good acceptance in both platforms to generate more income. 

To know which apps are our target we need to know which are the most common apps. With this information we will create frequency tables to visualize better the information.

The columns that could help us to find are:

Apple Store: `prime_genres` column index `11`

Google Play Store: `Genres` column index `9`

In [19]:
#Genera la tabla de frecuencias para cualquier dataset y cualquier columna
def freq_table(dataset, index):#index es para la columna
    longitud=len(dataset)
    diccionario={}
    for row in dataset:
        key=row[index]
        if key in diccionario:
            diccionario[key]+=1
        else:
            diccionario[key]=1
            
    for each in diccionario:
        value=diccionario[each]
        new=round((value/longitud)*100,2)
        diccionario[each]=new
    return diccionario
print(freq_table(final1,11))

{'Social Networking': 3.29, 'Photo & Video': 4.97, 'Games': 58.16, 'Music': 2.05, 'Reference': 0.56, 'Health & Fitness': 2.02, 'Weather': 0.87, 'Utilities': 2.51, 'Travel': 1.24, 'Shopping': 2.61, 'News': 1.33, 'Navigation': 0.19, 'Lifestyle': 1.58, 'Entertainment': 7.88, 'Food & Drink': 0.81, 'Sports': 2.14, 'Book': 0.43, 'Finance': 1.12, 'Education': 3.66, 'Productivity': 1.74, 'Business': 0.53, 'Catalogs': 0.12, 'Medical': 0.19}


In [20]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [21]:
#Tabla de frecuencias para iOs en la columna prime_genre
display_table(final1,11)

Games : 58.16
Entertainment : 7.88
Photo & Video : 4.97
Education : 3.66
Social Networking : 3.29
Shopping : 2.61
Utilities : 2.51
Sports : 2.14
Music : 2.05
Health & Fitness : 2.02
Productivity : 1.74
Lifestyle : 1.58
News : 1.33
Travel : 1.24
Finance : 1.12
Weather : 0.87
Food & Drink : 0.81
Reference : 0.56
Business : 0.53
Book : 0.43
Navigation : 0.19
Medical : 0.19
Catalogs : 0.12


In [22]:
#Tabla de frecuencias para Play Sore en la columna Category
display_table(final2,1)

FAMILY : 18.91
GAME : 9.72
TOOLS : 8.46
BUSINESS : 4.59
LIFESTYLE : 3.9
PRODUCTIVITY : 3.89
FINANCE : 3.7
MEDICAL : 3.53
SPORTS : 3.4
PERSONALIZATION : 3.32
COMMUNICATION : 3.24
HEALTH_AND_FITNESS : 3.08
PHOTOGRAPHY : 2.94
NEWS_AND_MAGAZINES : 2.8
SOCIAL : 2.66
TRAVEL_AND_LOCAL : 2.34
SHOPPING : 2.25
BOOKS_AND_REFERENCE : 2.14
DATING : 1.86
VIDEO_PLAYERS : 1.79
MAPS_AND_NAVIGATION : 1.4
FOOD_AND_DRINK : 1.24
EDUCATION : 1.16
ENTERTAINMENT : 0.96
LIBRARIES_AND_DEMO : 0.94
AUTO_AND_VEHICLES : 0.93
HOUSE_AND_HOME : 0.82
WEATHER : 0.8
EVENTS : 0.71
PARENTING : 0.65
ART_AND_DESIGN : 0.64
COMICS : 0.62
BEAUTY : 0.6


In [24]:
freqtable=freq_table(final1,11)
for k in freqtable:
    total=0#store the sum of user rating
    len_genre=0#number apps for each genre
    for each in final1:
        genre_app=each[11]
        if genre_app==k:
            count=float(each[5])
            total+=count
            len_genre+=1
    average=round(((total/len_genre)*100),2)
    print(k, ':', average)
        

Social Networking : 7154834.91
Photo & Video : 2844154.38
Games : 2278866.97
Music : 5732653.03
Reference : 7494211.11
Health & Fitness : 2329801.54
Weather : 5227989.29
Utilities : 1868445.68
Travel : 2824380.0
Shopping : 2691969.05
News : 2124802.33
Navigation : 8609033.33
Lifestyle : 1648576.47
Entertainment : 1402983.07
Food & Drink : 3333392.31
Sports : 2300889.86
Book : 3975850.0
Finance : 3146794.44
Education : 700398.31
Productivity : 2102841.07
Business : 749111.76
Catalogs : 400400.0
Medical : 61200.0


In [25]:
freqtable2=freq_table(final2,1)
for k in freqtable2:
    total2=0#store the sum of user rating
    len_category2=0#number apps for each genre
    for each in final2:
        category_app=each[1]
        if category_app==k:
            numero=each[5]
            numero=numero.replace('+','')
            numero=numero.replace(',','')
            numero=float(numero)
            total2+=numero
            len_category2+=1
    avg=round(((total2/len_category2)*100),2)
    print(k, ':', avg)

ART_AND_DESIGN : 198633508.77
AUTO_AND_VEHICLES : 64731781.71
BEAUTY : 51315188.68
BOOKS_AND_REFERENCE : 876781189.47
BUSINESS : 171229014.74
COMICS : 81765727.27
COMMUNICATION : 3845611916.72
DATING : 85402883.03
EDUCATION : 183349514.56
ENTERTAINMENT : 1164070588.24
EVENTS : 25354222.22
FINANCE : 138769247.56
FOOD_AND_DRINK : 192489773.64
HEALTH_AND_FITNESS : 418882198.53
HOUSE_AND_HOME : 133154056.16
LIBRARIES_AND_DEMO : 63850373.49
LIFESTYLE : 143781626.88
GAME : 1558801560.32
FAMILY : 369564181.98
MEDICAL : 12055061.98
SOCIAL : 2325365212.71
SHOPPING : 703687731.16
PHOTOGRAPHY : 1784011040.23
SPORTS : 363864014.29
TRAVEL_AND_LOCAL : 1398407771.01
TOOLS : 1080139129.87
PERSONALIZATION : 520148261.22
PRODUCTIVITY : 1678733134.49
PARENTING : 54260362.07
WEATHER : 507448619.72
VIDEO_PLAYERS : 2472787245.28
NEWS_AND_MAGAZINES : 954917846.77
MAPS_AND_NAVIGATION : 405694177.42


Analizing the previous information we can recomend to develope an app, based on the total apps genres downloads, in the genre `TRAVEL_AND_LOCAL` due the total of downloads of all the apps inside that genre are 1,398,407,771